File: v3/src/overlays/tracker-setup.ts
import type { OverlayManager, OverlayItem } from '../overlay.js';
import type { Config } from '../config.js';

interface LinearTeam {
  id: string;
  key: string;
  name: string;
}

interface LinearProject {
  id: string;
  name: string;
}

interface LinearWorkflowState {
  id: string;
  name: string;
  type: string;
}

async function linearGql(apiKey: string, query: string, variables: Record<string, unknown> = {}): Promise<any> {
  const res = await fetch('https://api.linear.app/graphql', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: apiKey },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) throw new Error(`Linear API ${res.status}`);
  const json = await res.json();
  if (json.errors?.length) throw new Error(json.errors[0].message);
  return json.data;
}

export function showTrackerSetup(
  overlay: OverlayManager,
  config: Config,
  onConfigured: () => void,
  onUpdate: () => void,
): void {
  overlay.show({
    title: 'Configure Issue Tracker',
    items: [
      { label: 'Select tracker type:', dimmed: true },
      { label: '', dimmed: true },
      {
        label: '[L] Linear',
        shortcut: 'l',
        action: () => setupLinear(overlay, config, onConfigured, onUpdate),
      },
      {
        label: '[J] Jira',
        shortcut: 'j',
        action: () => setupJira(overlay, config, onConfigured, onUpdate),
      },
      { label: '', dimmed: true },
      {
        label: '[N] No tracker',
        shortcut: 'n',
        action: () => {
          config.save({ tracker: { type: 'none' } });
          onConfigured();
          overlay.showMessage('Tracker disabled');
        },
      },
    ],
    onClose: onUpdate,
    footer: '[L] Linear  [J] Jira  [N] None',
  });
}

async function setupLinear(
  overlay: OverlayManager,
  config: Config,
  onConfigured: () => void,
  onUpdate: () => void,
): Promise<void> {
  // Check for API key
  const apiKey = process.env.LINEAR_API_KEY;
  if (!apiKey) {
    overlay.updateItems([
      { label: 'LINEAR_API_KEY not found in environment', dimmed: true },
      { label: '', dimmed: true },
      { label: 'To set up Linear:', dimmed: true },
      { label: '  1. Get an API key from Linear Settings > API', dimmed: true },
      { label: '  2. Add to your .env file:', dimmed: true },
      { label: '     LINEAR_API_KEY=lin_api_xxxxx', dimmed: true },
      { label: '  3. Restart orch3', dimmed: true },
      { label: '', dimmed: true },
      {
        label: '\u2190 Back',
        shortcut: 'b',
        action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
      },
    ]);
    return;
  }

  // Discover teams
  overlay.setLoading(true);
  try {
    const data = await linearGql(apiKey, `query {
      teams { nodes { id key name } }
    }`);
    const teams: LinearTeam[] = data.teams?.nodes || [];
    overlay.setLoading(false);

    if (teams.length === 0) {
      overlay.updateItems([
        { label: 'No teams found in your Linear workspace', dimmed: true },
        { label: '', dimmed: true },
        {
          label: '\u2190 Back',
          shortcut: 'b',
          action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
        },
      ]);
      return;
    }

    // Let user pick a team
    const items: OverlayItem[] = [
      { label: 'Select your Linear team:', dimmed: true },
      { label: '', dimmed: true },
    ];

    for (const team of teams) {
      items.push({
        label: `${team.key} — ${team.name}`,
        action: () => pickLinearProject(overlay, config, apiKey, team, onConfigured, onUpdate),
      });
    }

    items.push({ label: '', dimmed: true });
    items.push({
      label: '\u2190 Back',
      shortcut: 'b',
      action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
    });

    overlay.updateItems(items);
  } catch (err: any) {
    overlay.setLoading(false);
    overlay.updateItems([
      { label: `Error connecting to Linear: ${err.message}`, dimmed: true },
      { label: '', dimmed: true },
      {
        label: '[R] Retry',
        shortcut: 'r',
        action: () => setupLinear(overlay, config, onConfigured, onUpdate),
      },
      {
        label: '\u2190 Back',
        shortcut: 'b',
        action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
      },
    ]);
  }
}

async function pickLinearProject(
  overlay: OverlayManager,
  config: Config,
  apiKey: string,
  team: LinearTeam,
  onConfigured: () => void,
  onUpdate: () => void,
): Promise<void> {
  overlay.setLoading(true);
  try {
    // Fetch projects and workflow states for this team
    const data = await linearGql(apiKey, `query($teamId: String!) {
      team(id: $teamId) {
        projects { nodes { id name } }
        states { nodes { id name type } }
      }
    }`, { teamId: team.id });

    const projects: LinearProject[] = data.team?.projects?.nodes || [];
    const states: LinearWorkflowState[] = data.team?.states?.nodes || [];
    overlay.setLoading(false);

    // Build workflow mapping from state types
    const workflow: Record<string, string> = {};
    for (const state of states) {
      switch (state.type) {
        case 'backlog': workflow.queued = state.name; break;
        case 'started': workflow.in_progress = state.name; break;
        case 'completed': workflow.done = state.name; break;
        case 'cancelled': workflow.abandoned = state.name; break;
      }
    }

    const items: OverlayItem[] = [
      { label: `Team: ${team.key} — ${team.name}`, dimmed: true },
      { label: '', dimmed: true },
    ];

    if (projects.length > 0) {
      items.push({ label: 'Link to a Linear project (optional):', dimmed: true });
      items.push({ label: '', dimmed: true });

      // Option to skip project
      items.push({
        label: 'No project (team-level issues only)',
        action: () => saveLinearConfig(config, team, null, workflow, onConfigured, overlay),
      });

      for (const project of projects) {
        items.push({
          label: project.name,
          action: () => saveLinearConfig(config, team, project, workflow, onConfigured, overlay),
        });
      }
    } else {
      items.push({
        label: 'Save configuration',
        action: () => saveLinearConfig(config, team, null, workflow, onConfigured, overlay),
      });
    }

    items.push({ label: '', dimmed: true });
    items.push({
      label: '\u2190 Back to teams',
      shortcut: 'b',
      action: () => setupLinear(overlay, config, onConfigured, onUpdate),
    });

    overlay.updateItems(items);
  } catch (err: any) {
    overlay.setLoading(false);
    overlay.showMessage(`Error: ${err.message}`);
  }
}

function saveLinearConfig(
  config: Config,
  team: LinearTeam,
  project: LinearProject | null,
  workflow: Record<string, string>,
  onConfigured: () => void,
  overlay: OverlayManager,
): void {
  config.save({
    tracker: { type: 'linear', title_prefix: '[AI]' },
    linear: {
      team_key: team.key,
      ...(project ? { project_name: project.name } : {}),
      workflow,
      auth: { api_key_env: 'LINEAR_API_KEY' },
    },
  });

  onConfigured();

  const msg = project
    ? `Configured: ${team.key} / ${project.name}`
    : `Configured: ${team.key}`;
  overlay.showMessage(msg);

  // Show summary
  setTimeout(() => {
    overlay.updateItems([
      { label: `\u2713 Linear configured!`, dimmed: true },
      { label: '', dimmed: true },
      { label: `Team: ${team.key} — ${team.name}`, dimmed: true },
      ...(project ? [{ label: `Project: ${project.name}`, dimmed: true }] : []),
      { label: `Config saved to .orch/project.json`, dimmed: true },
      { label: '', dimmed: true },
      { label: 'Press Esc to close, then F3 to browse issues' , dimmed: true },
    ]);
  }, 500);
}

function setupJira(
  overlay: OverlayManager,
  config: Config,
  onConfigured: () => void,
  onUpdate: () => void,
): void {
  const baseUrl = process.env.JIRA_BASE_URL;
  const email = process.env.JIRA_EMAIL;
  const token = process.env.JIRA_API_TOKEN;

  if (!baseUrl || !email || !token) {
    const missing = [];
    if (!baseUrl) missing.push('JIRA_BASE_URL');
    if (!email) missing.push('JIRA_EMAIL');
    if (!token) missing.push('JIRA_API_TOKEN');

    overlay.updateItems([
      { label: 'Missing Jira environment variables:', dimmed: true },
      { label: '', dimmed: true },
      ...missing.map(v => ({ label: `  \u2717 ${v}`, dimmed: true })),
      { label: '', dimmed: true },
      { label: 'Add these to your .env file and restart orch3', dimmed: true },
      { label: '', dimmed: true },
      {
        label: '\u2190 Back',
        shortcut: 'b',
        action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
      },
    ]);
    return;
  }

  // Discover Jira projects
  discoverJiraProjects(overlay, config, baseUrl, email, token, onConfigured, onUpdate);
}

async function discoverJiraProjects(
  overlay: OverlayManager,
  config: Config,
  baseUrl: string,
  email: string,
  token: string,
  onConfigured: () => void,
  onUpdate: () => void,
): Promise<void> {
  overlay.setLoading(true);
  try {
    const auth = Buffer.from(`${email}:${token}`).toString('base64');
    const res = await fetch(`${baseUrl}/rest/api/3/project/search?maxResults=50`, {
      headers: { Authorization: `Basic ${auth}`, Accept: 'application/json' },
    });
    if (!res.ok) throw new Error(`Jira API ${res.status}`);
    const data = await res.json();
    const projects = data.values || [];
    overlay.setLoading(false);

    if (projects.length === 0) {
      overlay.updateItems([
        { label: 'No projects found in Jira', dimmed: true },
        { label: '', dimmed: true },
        {
          label: '\u2190 Back',
          shortcut: 'b',
          action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
        },
      ]);
      return;
    }

    const items: OverlayItem[] = [
      { label: 'Select your Jira project:', dimmed: true },
      { label: '', dimmed: true },
    ];

    for (const proj of projects) {
      items.push({
        label: `${proj.key} — ${proj.name}`,
        action: () => {
          config.save({
            tracker: { type: 'jira', title_prefix: '[AI]' },
            jira: {
              project_key: proj.key,
              issue_type: 'Task',
              subtask_issue_type: 'Sub-task',
              status: { todo: 'To Do', in_progress: 'In Progress', done: 'Done' },
              auth: {
                base_url_env: 'JIRA_BASE_URL',
                email_env: 'JIRA_EMAIL',
                api_token_env: 'JIRA_API_TOKEN',
              },
            },
          });
          onConfigured();
          overlay.showMessage(`Configured: ${proj.key}`);
          setTimeout(() => {
            overlay.updateItems([
              { label: `\u2713 Jira configured!`, dimmed: true },
              { label: '', dimmed: true },
              { label: `Project: ${proj.key} — ${proj.name}`, dimmed: true },
              { label: `Config saved to .orch/project.json`, dimmed: true },
              { label: '', dimmed: true },
              { label: 'Press Esc to close, then F3 to browse issues', dimmed: true },
            ]);
          }, 500);
        },
      });
    }

    items.push({ label: '', dimmed: true });
    items.push({
      label: '\u2190 Back',
      shortcut: 'b',
      action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
    });

    overlay.updateItems(items);
  } catch (err: any) {
    overlay.setLoading(false);
    overlay.updateItems([
      { label: `Error connecting to Jira: ${err.message}`, dimmed: true },
      { label: '', dimmed: true },
      {
        label: '[R] Retry',
        shortcut: 'r',
        action: () => discoverJiraProjects(overlay, config, baseUrl, email, token, onConfigured, onUpdate),
      },
      {
        label: '\u2190 Back',
        shortcut: 'b',
        action: () => showTrackerSetup(overlay, config, onConfigured, onUpdate),
      },
    ]);
  }
}

File: app-native/Sources/Orch/Overlays/TrackerSetup.swift
import Foundation

func showTrackerSetup(overlay: OverlayManager, config: Config, onConfigured: @escaping () -> Void, onUpdate: @escaping () -> Void) {
    overlay.show(OverlayConfig(
        title: "Configure Issue Tracker",
        items: [
            OverlayItem(label: "Select tracker type:", dimmed: true),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "[L] Linear", shortcut: "l", action: {
                setupLinear(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
            }),
            OverlayItem(label: "[J] Jira", shortcut: "j", action: {
                setupJira(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
            }),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "[N] No tracker", shortcut: "n", action: {
                config.save(updates: ProjectJson(tracker: TrackerConfig(type: "none")))
                onConfigured()
                overlay.showMessage("Tracker disabled")
            }),
        ],
        onClose: onUpdate,
        footer: "[L] Linear  [J] Jira  [N] None"
    ))
}

private func setupLinear(overlay: OverlayManager, config: Config, onConfigured: @escaping () -> Void, onUpdate: @escaping () -> Void) {
    guard let apiKey = ProcessInfo.processInfo.environment["LINEAR_API_KEY"] else {
        overlay.updateItems([
            OverlayItem(label: "LINEAR_API_KEY not found in environment", dimmed: true),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "To set up Linear:", dimmed: true),
            OverlayItem(label: "  1. Get an API key from Linear Settings > API", dimmed: true),
            OverlayItem(label: "  2. Add to your .env file:", dimmed: true),
            OverlayItem(label: "     LINEAR_API_KEY=lin_api_xxxxx", dimmed: true),
            OverlayItem(label: "  3. Restart orch", dimmed: true),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
            }),
        ])
        return
    }

    overlay.setLoading(true)
    Task {
        do {
            let teams = try await fetchLinearTeams(apiKey: apiKey)
            await MainActor.run {
                overlay.setLoading(false)
                if teams.isEmpty {
                    overlay.updateItems([
                        OverlayItem(label: "No teams found in your Linear workspace", dimmed: true),
                        OverlayItem(label: "", dimmed: true),
                        OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                            showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                        }),
                    ])
                    return
                }

                var items: [OverlayItem] = [
                    OverlayItem(label: "Select your Linear team:", dimmed: true),
                    OverlayItem(label: "", dimmed: true),
                ]
                for team in teams {
                    items.append(OverlayItem(label: "\(team.key) \u{2014} \(team.name)", action: {
                        pickLinearProject(overlay: overlay, config: config, apiKey: apiKey, team: team, onConfigured: onConfigured, onUpdate: onUpdate)
                    }))
                }
                items.append(OverlayItem(label: "", dimmed: true))
                items.append(OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                    showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                }))
                overlay.updateItems(items)
            }
        } catch {
            await MainActor.run {
                overlay.setLoading(false)
                overlay.updateItems([
                    OverlayItem(label: "Error connecting to Linear: \(error.localizedDescription)", dimmed: true),
                    OverlayItem(label: "", dimmed: true),
                    OverlayItem(label: "[R] Retry", shortcut: "r", action: {
                        setupLinear(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                    }),
                    OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                        showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                    }),
                ])
            }
        }
    }
}

private struct LinearTeamInfo { let id: String; let key: String; let name: String }
private struct LinearProjectInfo { let id: String; let name: String }

private func fetchLinearTeams(apiKey: String) async throws -> [LinearTeamInfo] {
    let data = try await linearGql(apiKey: apiKey, query: "query { teams { nodes { id key name } } }")
    guard let teams = data["teams"] as? [String: Any],
          let nodes = teams["nodes"] as? [[String: Any]] else { return [] }
    return nodes.map { LinearTeamInfo(id: $0["id"] as? String ?? "", key: $0["key"] as? String ?? "", name: $0["name"] as? String ?? "") }
}

private func pickLinearProject(overlay: OverlayManager, config: Config, apiKey: String, team: LinearTeamInfo, onConfigured: @escaping () -> Void, onUpdate: @escaping () -> Void) {
    overlay.setLoading(true)
    Task {
        do {
            let data = try await linearGql(apiKey: apiKey, query: """
            query($teamId: String!) {
              team(id: $teamId) {
                projects { nodes { id name } }
                states { nodes { id name type } }
              }
            }
            """, variables: ["teamId": team.id])

            let teamData = data["team"] as? [String: Any] ?? [:]
            let projects = ((teamData["projects"] as? [String: Any])?["nodes"] as? [[String: Any]] ?? []).map {
                LinearProjectInfo(id: $0["id"] as? String ?? "", name: $0["name"] as? String ?? "")
            }
            let states = ((teamData["states"] as? [String: Any])?["nodes"] as? [[String: Any]] ?? [])

            var workflow: [String: String] = [:]
            for state in states {
                switch state["type"] as? String {
                case "backlog": workflow["queued"] = state["name"] as? String
                case "started": workflow["in_progress"] = state["name"] as? String
                case "completed": workflow["done"] = state["name"] as? String
                case "cancelled": workflow["abandoned"] = state["name"] as? String
                default: break
                }
            }

            await MainActor.run {
                overlay.setLoading(false)
                var items: [OverlayItem] = [
                    OverlayItem(label: "Team: \(team.key) \u{2014} \(team.name)", dimmed: true),
                    OverlayItem(label: "", dimmed: true),
                ]

                if !projects.isEmpty {
                    items.append(OverlayItem(label: "Link to a Linear project (optional):", dimmed: true))
                    items.append(OverlayItem(label: "", dimmed: true))
                    items.append(OverlayItem(label: "No project (team-level issues only)", action: {
                        saveLinearConfig(config: config, team: team, project: nil, workflow: workflow, onConfigured: onConfigured, overlay: overlay)
                    }))
                    for project in projects {
                        items.append(OverlayItem(label: project.name, action: {
                            saveLinearConfig(config: config, team: team, project: project, workflow: workflow, onConfigured: onConfigured, overlay: overlay)
                        }))
                    }
                } else {
                    items.append(OverlayItem(label: "Save configuration", action: {
                        saveLinearConfig(config: config, team: team, project: nil, workflow: workflow, onConfigured: onConfigured, overlay: overlay)
                    }))
                }

                items.append(OverlayItem(label: "", dimmed: true))
                items.append(OverlayItem(label: "\u{2190} Back to teams", shortcut: "b", action: {
                    setupLinear(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                }))
                overlay.updateItems(items)
            }
        } catch {
            await MainActor.run {
                overlay.setLoading(false)
                overlay.showMessage("Error: \(error.localizedDescription)")
            }
        }
    }
}

private func saveLinearConfig(config: Config, team: LinearTeamInfo, project: LinearProjectInfo?, workflow: [String: String], onConfigured: @escaping () -> Void, overlay: OverlayManager) {
    config.save(updates: ProjectJson(
        tracker: TrackerConfig(type: "linear", title_prefix: "[AI]"),
        linear: LinearConfig(
            team_key: team.key,
            project_name: project?.name,
            workflow: workflow,
            auth: LinearConfig.LinearAuth(api_key_env: "LINEAR_API_KEY")
        )
    ))

    onConfigured()
    let msg = project != nil ? "Configured: \(team.key) / \(project!.name)" : "Configured: \(team.key)"
    overlay.showMessage(msg)

    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
        var items: [OverlayItem] = [
            OverlayItem(label: "\u{2713} Linear configured!", dimmed: true),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "Team: \(team.key) \u{2014} \(team.name)", dimmed: true),
        ]
        if let p = project {
            items.append(OverlayItem(label: "Project: \(p.name)", dimmed: true))
        }
        items += [
            OverlayItem(label: "Config saved to .orch/project.json", dimmed: true),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "Press Esc to close", dimmed: true),
        ]
        overlay.updateItems(items)
    }
}

private func setupJira(overlay: OverlayManager, config: Config, onConfigured: @escaping () -> Void, onUpdate: @escaping () -> Void) {
    let env = ProcessInfo.processInfo.environment
    let baseUrl = env["JIRA_BASE_URL"]
    let email = env["JIRA_EMAIL"]
    let token = env["JIRA_API_TOKEN"]

    guard let baseUrl = baseUrl, let email = email, let token = token else {
        var missing: [String] = []
        if baseUrl == nil { missing.append("JIRA_BASE_URL") }
        if email == nil { missing.append("JIRA_EMAIL") }
        if token == nil { missing.append("JIRA_API_TOKEN") }

        var items: [OverlayItem] = [OverlayItem(label: "Missing Jira environment variables:", dimmed: true), OverlayItem(label: "", dimmed: true)]
        for v in missing { items.append(OverlayItem(label: "  \u{2717} \(v)", dimmed: true)) }
        items += [
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "Add these to your .env file and restart orch", dimmed: true),
            OverlayItem(label: "", dimmed: true),
            OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
            }),
        ]
        overlay.updateItems(items)
        return
    }

    overlay.setLoading(true)
    Task {
        do {
            let auth = Data("\(email):\(token)".utf8).base64EncodedString()
            let url = URL(string: "\(baseUrl)/rest/api/3/project/search?maxResults=50")!
            var request = URLRequest(url: url)
            request.setValue("Basic \(auth)", forHTTPHeaderField: "Authorization")
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            let (data, _) = try await URLSession.shared.data(for: request)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let projects = json["values"] as? [[String: Any]] else { return }

            await MainActor.run {
                overlay.setLoading(false)
                if projects.isEmpty {
                    overlay.updateItems([
                        OverlayItem(label: "No projects found in Jira", dimmed: true),
                        OverlayItem(label: "", dimmed: true),
                        OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                            showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                        }),
                    ])
                    return
                }

                var items: [OverlayItem] = [
                    OverlayItem(label: "Select your Jira project:", dimmed: true),
                    OverlayItem(label: "", dimmed: true),
                ]
                for proj in projects {
                    let key = proj["key"] as? String ?? ""
                    let name = proj["name"] as? String ?? ""
                    items.append(OverlayItem(label: "\(key) \u{2014} \(name)", action: {
                        config.save(updates: ProjectJson(
                            tracker: TrackerConfig(type: "jira", title_prefix: "[AI]"),
                            jira: JiraConfig(
                                project_key: key, issue_type: "Task", subtask_issue_type: "Sub-task",
                                status: ["todo": "To Do", "in_progress": "In Progress", "done": "Done"],
                                auth: JiraConfig.JiraAuth(base_url_env: "JIRA_BASE_URL", email_env: "JIRA_EMAIL", api_token_env: "JIRA_API_TOKEN")
                            )
                        ))
                        onConfigured()
                        overlay.showMessage("Configured: \(key)")
                        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                            overlay.updateItems([
                                OverlayItem(label: "\u{2713} Jira configured!", dimmed: true),
                                OverlayItem(label: "", dimmed: true),
                                OverlayItem(label: "Project: \(key) \u{2014} \(name)", dimmed: true),
                                OverlayItem(label: "Config saved to .orch/project.json", dimmed: true),
                                OverlayItem(label: "", dimmed: true),
                                OverlayItem(label: "Press Esc to close", dimmed: true),
                            ])
                        }
                    }))
                }
                items.append(OverlayItem(label: "", dimmed: true))
                items.append(OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                    showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                }))
                overlay.updateItems(items)
            }
        } catch {
            await MainActor.run {
                overlay.setLoading(false)
                overlay.updateItems([
                    OverlayItem(label: "Error connecting to Jira: \(error.localizedDescription)", dimmed: true),
                    OverlayItem(label: "", dimmed: true),
                    OverlayItem(label: "[R] Retry", shortcut: "r", action: {
                        setupJira(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                    }),
                    OverlayItem(label: "\u{2190} Back", shortcut: "b", action: {
                        showTrackerSetup(overlay: overlay, config: config, onConfigured: onConfigured, onUpdate: onUpdate)
                    }),
                ])
            }
        }
    }
}

private func linearGql(apiKey: String, query: String, variables: [String: Any] = [:]) async throws -> [String: Any] {
    let url = URL(string: "https://api.linear.app/graphql")!
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue(apiKey, forHTTPHeaderField: "Authorization")
    request.httpBody = try JSONSerialization.data(withJSONObject: ["query": query, "variables": variables])

    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
        throw NSError(domain: "Linear", code: (response as? HTTPURLResponse)?.statusCode ?? 0)
    }
    guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw NSError(domain: "Linear", code: 0)
    }
    if let errors = json["errors"] as? [[String: Any]], let msg = errors.first?["message"] as? String {
        throw NSError(domain: "Linear", code: 0, userInfo: [NSLocalizedDescriptionKey: msg])
    }
    return json["data"] as? [String: Any] ?? [:]
}
